ITE/4/README.adoc

```diff
@@ -0,0 +1,361 @@
+
+| Sponsor
+| link:https://github.com/santiagotorres[Santiago Torres-Arias]
+
+| Status
+| Draft :speech_balloon:
+
+| Type
+| Standards
+
+| Created
+| 2019-09-30
+
+|===
+
+[[abstract]]
+== Abstract
+
+This ITE proposes a generic Unique Resource Identifier (URI) scheme for in-toto
+artifacts. Currently, the in-toto specification only allows for files in the
+local filesystem to be the artifacts of link metadata. This ITE also defines the
+scheme for a generic URI and provides a behaviour of in-toto implementations
+that support generic URIs.
+
+[[specification]]
+== Specification
+
+The current in-toto specification must be modified to provide support for the
+generic URI schemes. The current specification allows only for files in the
+local filesystem to be set as artifacts. With the addition of other entities,
+the parts of the specification that directly interact or handle the artifacts
+must be updated.
+
+in-toto link metadata files currently look like this:
+
+```
+{
+    "_type": "link",
+    "name": "<NAME>",
+    "command": "<COMMAND>",
+    "materials": {
+        "<PATH>": <HASH>,
+        "...": ...
+    },
+    "products": {
+        "<PATH>": <HASH>,
+        "...": ...
+    },
+    "byproducts": {
+        "stderr": "<STDERR>",
+        "stdout": "<STDOUT>",
+        "return-value": "<RETURN-VALUE>"
+    },
+    "environment": {
+        "variables": "<ENV>",
+        "filesystem": "<FS>",
+        "workdir": "<PWD>"
+    }
+}
+```
+
+The `materials` and `products` fields contain key-value pairs. The URI for each
+artifact is its path in the filesystem. With this ITE active, the artifact could
+instead have a generic URI which can be resolved to some external resource that
+is part of the supply chain. The URI MUST have a fixed structure and MUST
+contain all the information necessary for the verifier to resolve if any part of
+the verification workflow requires it.
+
+This ITE proposes the following structure for the URI.
+
+`<scheme>:<hier-part>`
+
+This structure is derived from RFC 3986. The `scheme` defines the token or
+identifier for the type of resource the URI refers to. It can also contain
+other relevant information that indicates how the resource can be accessed. The
+`hier-part` identifies the path or location of the resource.
+
+A compliant in-toto system must be capable of generating link metadata with
+artifacts that have generic URI references. Therefore, this system must also be
+capable of resolving this URI and calculating cryptographic hashes of these
+artifacts. It's important to note that the contents of the resource are hashed,
+as opposed to the URI itself.
+
+To that end, it's important for a URI scheme to have a single mapping for the
+service it refers to. Consider an example of a GitHub specific resource - a pull
+request or an issue.
+
+`github+https://github.com/in-toto/in-toto/pull/250`
+
+This can also instead use GitHub's API specific URI to refer to the same data.
```

Comment:
What does "same data" mean? Speaking in abstract terms `github+https://github.com/in-toto/in-toto/pull/250` and `github+https://api.github.com/repos/in-toto/in-toto/pulls/250` point to the same *thing*. But how do those *things* get hashed?
```
`curl https://github.com/in-toto/in-toto/pull/250 | shasum -a 256` != `curl https://api.github.com/repos/in-toto/in-toto/pulls/250 | shasum -a 256`
```